clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -183,22 +183,33 @@ public RecordAppendResult append(TopicPartition tp,
             ByteBuffer buffer = free.allocate(size, maxTimeToBlock);
             synchronized (dq) {
                 // Need to check if producer is closed again after grabbing the dequeue lock.
-                if (closed)
+                if (closed) {
+                    free.deallocate(buffer);
                     throw new IllegalStateException("Cannot send after the producer is closed.");
+                }
+
+                RecordAppendResult appendResult = null;
+                try {
+                    appendResult = tryAppend(timestamp, key, value, callback, dq);
+                } catch (Exception e) {
```

Comment:
@ijuma I am not sure whether we are trying to deallocate buffer after it is used as this is internally wrapped by `MemoryRecordsBuilder#bufferStream`. My understanding was this buffer should be deallocated only when the buffer is not effectively used in building the record and appending it to the batch.